### PR TITLE
MA-86 Refactor coupon verification to return responses only in endpoint handler

### PR DESCRIPTION
**Pull request by njfaries**
# Description
Continuation of work started in https://github.com/BoltApp/bolt-magento2/pull/949 to move error responses to the endpoints and throw exception from internal methods.

Fixes: MA-86

#changelog [WIP] MA-86 Refactor coupon verification to return responses only in endpoint handler

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Comment by njfaries** [comment]
The test `removeDiscount_removeGiftCard()` was failing silently before. After the changes introduced by this PR I noticed that it was never properly finishing the testing the method it was supposed to test; it was calling the `verifyCoupon` method I changed and originally returning explicitly false instead of throwing an exception. Now that `verifyCoupon` throws an exception the method under test, `removeDiscount`, was returning false from the `catch` statement when before it was just returning true. I updated the test to actually have a giftcard mock to remove.

**Review by danac-gs — CHANGES_REQUESTED** [review]
You should also catch the specific BoltException here https://github.com/BoltApp/bolt-magento2/blob/master/Model/Api/UpdateDiscountTrait.php#L470

**Comment by njfaries** [comment]
> You should also catch the specific BoltException here https://github.com/BoltApp/bolt-magento2/blob/master/Model/Api/UpdateDiscountTrait.php#L470

I believe this should actually be more similar to the response sending that I removed, right? I think the `removeDiscount` method should throw or rethrow a `boltException` to be handled in the `UpdateCart` class, in the same way the exception thrown by `verifyCouponCode` is being handled.

**Comment by njfaries** [comment]
This has recursed into quite the PR. I've been changing helper functions to not send responses but throw `BoltExcepctions` instead, and have been hunting down other uses to update them to expect the exceptions as well. Tests still have yet to be updated. I believe this is generally the right track though to be on.

**Comment by njfaries** [comment]
I've been making my way through trying to narrow the scope to just what is strictly necessary for the `validate` method in `DiscountCodeValidation.php`, while making sure that any methods that now throw exceptions that are used elsewhere still have the exceptions caught. In general they should, as the overlap is with other endpoint handlers (primarily the `UpdateCart` one), which have everything wrapped up in a large try-catch. I've only had to make a few minor tweaks to unit tests, specifically just the ones testing internal methods as they changed from expecting a response to expecting an exception. The handler tests all still pass, so I believe this is sound.  I'll be going through the checkout/discount flow manually in the morning to make sure and hopefully removing the WIP tag.

File: Test/Unit/Model/Api/UpdateDiscountTraitTest.php [hunk 1]
```diff
@@ -976,42 +976,39 @@ public function applyingCouponCode_virtualImmutableQuote()
...
+
+    //     $quote = $this->getQuoteMock();
+
+    //     $exception = new \Exception('General exception');
+    //     $this->discountHelper->expects(self::once())->method('setCouponCode')
+    //         ->with($quote, self::COUPON_CODE)->willThrowException($exception);
+
+    //     $result = TestHelper::invokeMethod($this->currentMock, 'applyingCouponCode', [self::COUPON_CODE, $coupon, $quote]);
+
+    //     $this->assertFalse($result);
```
Comment (danac-gs) [inline Test/Unit/Model/Api/UpdateDiscountTraitTest.php:1010]: remove commented code if redundant

**Review by danac-gs — APPROVED** [review]
lgtm

**Review by danac-gs — DISMISSED** [review]

File: Model/Api/UpdateDiscountTrait.php [hunk 2]
```diff
@@ -297,70 +281,49 @@ protected function applyingCouponCode($couponCode, $coupon, $quote, $addQuote =
...
             if (!in_array(0, $groupIds)) {
                 $this->logHelper->addInfoLog('Error: coupon requires login.');
-                $this->sendErrorResponse(
-                    BoltErrorResponse::ERR_CODE_REQUIRES_LOGIN,
-                    sprintf('The coupon code %s requires login', $couponCode),
-                    422,
-                    $quote
-                );
-                return false;
+                    throw new BoltException(
```
Comment (vitaliyreznikov) [inline Model/Api/UpdateDiscountTrait.php:307]: Do we need to fix indents here?

**Review by vitaliyreznikov — APPROVED** [review]
LGTM

File: Model/Api/UpdateDiscountTrait.php (hunk 2 above)
Comment (njfaries) [reply Model/Api/UpdateDiscountTrait.php:307]: Good catch, fixing

**Review by vitaliyreznikov — APPROVED** [review]

**Pull request merged**